crates/sui-light-client/src/checkpoint.rs

```diff
@@ -0,0 +1,275 @@
+use std::io::Read;
+use std::sync::Arc;
+use std::{fs, io::Write};
+use sui_archival::read_manifest;
+use sui_config::genesis::Genesis;
+use sui_sdk::SuiClientBuilder;
+use sui_storage::object_store::http::HttpDownloaderBuilder;
+use sui_storage::object_store::ObjectStoreGetExt;
+use sui_types::committee::Committee;
+use sui_types::messages_checkpoint::EndOfEpochData;
+use sui_types::{
+    crypto::AuthorityQuorumSignInfo, message_envelope::Envelope,
+    messages_checkpoint::CheckpointSummary,
+};
+use tracing::info;
+use url::Url;
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct CheckpointsList {
+    pub checkpoints: Vec<u64>,
+}
+
+pub fn read_checkpoint_list(config: &Config) -> Result<CheckpointsList> {
+    let checkpoints_path = config.checkpoint_list_path();
+    let reader = fs::File::open(checkpoints_path)?;
+    Ok(serde_yaml::from_reader(reader)?)
+}
+
+pub fn write_checkpoint_list(config: &Config, checkpoints_list: &CheckpointsList) -> Result<()> {
+    let checkpoints_path = config.checkpoint_list_path();
+    let mut writer = fs::File::create(checkpoints_path)?;
+    let bytes = serde_yaml::to_vec(checkpoints_list)?;
+    writer
+        .write_all(&bytes)
+        .map_err(|e| anyhow!("Unable to serialize checkpoint list: {}", e))
+}
+
+pub fn read_checkpoint(
+    config: &Config,
+    seq: u64,
+) -> Result<Envelope<CheckpointSummary, AuthorityQuorumSignInfo<true>>> {
+    read_checkpoint_general(config, seq, None)
+}
+
+fn read_checkpoint_general(
+    config: &Config,
+    seq: u64,
+    path: Option<&str>,
+) -> Result<Envelope<CheckpointSummary, AuthorityQuorumSignInfo<true>>> {
+    let checkpoint_path = config.checkpoint_path(seq, path);
+    let mut reader = fs::File::open(checkpoint_path)?;
+    let mut buffer = Vec::new();
+    reader.read_to_end(&mut buffer)?;
+    bcs::from_bytes(&buffer).map_err(|_| anyhow!("Unable to parse checkpoint file"))
+}
+
+pub fn write_checkpoint(
+    config: &Config,
+    summary: &Envelope<CheckpointSummary, AuthorityQuorumSignInfo<true>>,
+) -> Result<()> {
+    write_checkpoint_general(config, summary, None)
+}
+
+fn write_checkpoint_general(
+    config: &Config,
+    summary: &Envelope<CheckpointSummary, AuthorityQuorumSignInfo<true>>,
+    path: Option<&str>,
+) -> Result<()> {
+    let checkpoint_path = config.checkpoint_path(*summary.sequence_number(), path);
+    let mut writer = fs::File::create(checkpoint_path)?;
+    let bytes =
+        bcs::to_bytes(summary).map_err(|_| anyhow!("Unable to serialize checkpoint summary"))?;
+    writer.write_all(&bytes)?;
+    Ok(())
+}
+
+/// Downloads the list of end of epoch checkpoints from the archive store or the GraphQL endpoint
+async fn sync_checkpoint_list_to_latest(config: &Config) -> anyhow::Result<()> {
+    if config.graphql_url.is_some() {
+        sync_checkpoint_list_to_latest_using_graphql(config).await
+    } else {
+        sync_checkpoint_list_to_latest_using_archive(config).await
+    }
+}
```

Comment:
I am tilted towards just syncing using both. Archive seems like the only option for historic sync, whereas GraphQL seems more up-to-date for latest ones (e.g., when I tried to sync on testnet, archive didn't have the latest two epochs).

```suggestion
async fn sync_checkpoint_list_to_latest(config: &Config) -> anyhow::Result<()> {
    if config.archive_store_config.is_none() && config.graphql_url.is_none() {
        return Err(anyhow!("Archive store config or GraphQL URL is not provided"));
    }

    info!("Syncing checkpoint list to latest using archive store and GraphQL");
    sync_checkpoint_list_to_latest_using_archive(config).await?;
    sync_checkpoint_list_to_latest_using_graphql(config).await?;
    Ok(())
}
```